File: 024_Swap_Nodes_in_Pairs/024_1.py
from typing import List, Dict
from queue import Queue
from collections import deque, Counter
import math
from heapq import heappush, heappushpop

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        next_ = head.next
        head.next = self.swapPairs(next_.next)
        next_.next = head
        return next_


if __name__ == "__main__":
    s = Solution()
    node1 = ListNode(1)
    node2 = ListNode(2)
    node3 = ListNode(3)
    node4 = ListNode(4)
    node1.next = node2
    node2.next = node3
    node3.next = node4
    nn = s.swapPairs(node1)
    while nn:
        print(nn.val, end=" ")
        nn = nn.next

File: 024_Swap_Nodes_in_Pairs/024_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <stack>
#include <bitset>
#include <array>
#include <queue>
#include <sstream>
#include <set>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x): val(x), next(NULL) { }
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head) return head;
        ListNode *dummy = new ListNode(0);
        ListNode *pre = dummy;
        pre->next = head;
        while (head && head->next) {
            ListNode *end = head->next->next;
            pre->next = head->next;
            head->next->next = head;
            head->next = end;
            pre = head;
            head = head->next;
        }
        return dummy->next;
    }
};

int main() {
    ListNode *node1 = new ListNode(1);
    ListNode *node2 = new ListNode(2);
    ListNode *node3 = new ListNode(3);
    ListNode *node4 = new ListNode(4);
    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    Solution s;
    ListNode *res = s.swapPairs(node1);
    while (res) {
        cout << res->val << " ";
        res = res->next;
    }
}

